runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/DoFnOperator.java

```diff
@@ -683,18 +721,25 @@ protected final void invokeFinishBundle() {
   }
 
   @Override
-  public void snapshotState(StateSnapshotContext context) throws Exception {
-
-    // Forced finish a bundle in checkpoint barrier otherwise may lose data.
-    // Careful, it use OperatorState or KeyGroupState to store outputs, So it
-    // must be called before their snapshot.
+  public final void snapshotState(StateSnapshotContext context) throws Exception {
+    // We can't output here anymore because the checkpoint barrier has already been
+    // sent downstream. This is going to change with 1.6/1.7's prepareSnapshotBarrier.
     outputManager.openBuffer();
     invokeFinishBundle();
     outputManager.closeBuffer();
 
     super.snapshotState(context);
   }
 
+  @Override
+  public final void notifyCheckpointComplete(long checkpointId) throws Exception {
+    super.notifyCheckpointComplete(checkpointId);
+    // We have to release all buffered data collected during checkpointing
```

Comment:
This is a bit misleading. Perhaps say instead "We have to release all data buffered till the checkpoint". Also, do we need to pay attention to the `checkpointId` here? Isn't it possible that  more than one checkpoint is in progress (this notification is async)?

Reply:
I think that is as much misleading as the current version. Perhaps "We can now release all buffered data which was held back for @RequiresStableInput guarantees."

We do not allow more than one checkpoint at a time. Consequently, this will be called before the next checkpoint completes.

Reply:
"This method is called as a notification once a distributed checkpoint has been completed."

https://github.com/apache/flink/blob/release-1.5/flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointListener.java#L40

Doesn't that mean that elements can be processed between `snapshotState` and `notifyCheckpointComplete`. And if so, we should not emit those elements since they are not part of the checkpoint?

Regarding the comment, all I can say is that it does not resonate. The elements are not collected "during" checkpointing, but rather between checkpoints.

Reply:
>Regarding the comment, all I can say is that it does not resonate. The elements are not collected "during" checkpointing, but rather between checkpoints.

What about the suggestion above to rephrase the comment to "We can now release all buffered data which was held back for @RequiresStableInput guarantees."?

>Doesn't that mean that elements can be processed between snapshotState and notifyCheckpointComplete. And if so, we should not emit those elements since they are not part of the checkpoint?

That's correct, we need to exclude the items which arrive after snapshotState has been called. We probably want to keep them in a separate state namespace.

Reply:
@tweise I've push an update to the `BufferingDoFnRunner` to open a new buffer for any elements which arrive after snapshotState. This now even supports multiple snapshots executing which is a setting we currently do not expose via FlinkPipelineOptions.

Reply:
> What about the suggestion above to rephrase the comment to "We can now release all buffered data which was held back for @RequiresStableInput guarantees."?

That would be clearer, and maybe be more specific about what buffered data we can release.